cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesDataset.java

```diff
@@ -0,0 +1,177 @@
+
+  private PreferencesDataset(ConfigDataset configDataset) {
+    this.configDataset = configDataset;
+  }
+
+  public static PreferencesDataset get(DatasetContext datasetContext, DatasetFramework dsFramework) {
+    return new PreferencesDataset(ConfigDataset.get(datasetContext, dsFramework));
+  }
+
+  private Map<String, String> getConfigProperties(String namespace, String id) {
+    Map<String, String> value = new HashMap<>();
+    try {
+      Config config = configDataset.get(namespace, PREFERENCES_CONFIG_TYPE, id);
+      value.putAll(config.getProperties());
+    } catch (ConfigNotFoundException e) {
+      //no-op - return empty map
+    }
+    return value;
+  }
+
+  private void setConfig(String namespace, String id, Map<String, String> propertyMap) {
+    Config config = new Config(id, propertyMap);
+    configDataset.createOrUpdate(namespace, PREFERENCES_CONFIG_TYPE, config);
+  }
+
+  private void deleteConfig(String namespace, String id) {
+    try {
+      configDataset.delete(namespace, PREFERENCES_CONFIG_TYPE, id);
+    } catch (ConfigNotFoundException e) {
+      //no-op
+    }
+  }
+
+  public Map<String, String> getProperties(EntityId entityId) {
+    switch (entityId.getEntityType()) {
+      case INSTANCE:
+        return getConfigProperties(EMPTY_NAMESPACE, getMultipartKey(INSTANCE_PROPERTIES));
+      case NAMESPACE:
+        NamespaceId namespaceId = (NamespaceId) entityId;
+        return getConfigProperties(namespaceId.getNamespace(), getMultipartKey(namespaceId.getNamespace()));
+      case APPLICATION:
+        ApplicationId appId = (ApplicationId) entityId;
+        return getConfigProperties(appId.getNamespace(), getMultipartKey(appId.getNamespace(), appId.getApplication()));
+      case PROGRAM:
+        ProgramId programId = (ProgramId) entityId;
+        return getConfigProperties(programId.getNamespace(),
+                                   getMultipartKey(programId.getNamespace(), programId.getApplication(),
+                                                   programId.getType().getCategoryName(), programId.getProgram()));
+      default:
+        return new HashMap<>();
+    }
+  }
+
+  public Map<String, String> getResolvedProperties(EntityId entityId) {
+    // if it is instance level get the properties and return
+    if (entityId.getEntityType().equals(EntityType.INSTANCE)) {
+      return getProperties(entityId);
+    }
+
+    // get properties from current level
+    Map<String, String> properties = getProperties(entityId);
+    // if the entity id has a parent id, get the preference from its parent
+    if (entityId instanceof ParentedId) {
+      properties.putAll(getProperties(((ParentedId) entityId).getParent()));
+    } else {
+      // otherwise it is a namespace id, which we want to look at the instance level
+      properties.putAll(getProperties(new InstanceId("")));
+    }
+    return properties;
+  }
+
+  public void setProperties(EntityId entityId, Map<String, String> propMap) {
```

Comment:
Please add javadoc for at least all public methods. For non-public methods, it is also good to have javadoc to state the method contract if it is non-trivial.

Reply:
Added